draft-dekater-scion-dataplane.md

```diff
@@ -1855,22 +1855,32 @@ Tunneling IP traffic over SCION requires a pair of SIGs and it involves the foll
+IP packets are encapsulated over SCION/UDP into SIG frames. Whilst in principle, a given pair of SIGs may use any tunneling protocol, existing deployments use the SIG framing as described here. This protocol is designed to:
+
+- provide independence from the underlying SCION path MTU which can increase and decrease over time.
+- provide fast detection of packet loss and subsequent recovery of decapsulation for packets that weren't lost.
+- support for multiple streams within a framing session such that independent packet sequences be tunneled in parallel.
 
 There may be multiple IP packets in a single SIG frame, and a single IP packet may be split into multiple SIG frames.
-A source SIG unidirectionally establishes a SIG tunneling session with a destination SIG. A session may be mapped to one or multiple SCION paths, depending on implementors.
 
-Streams can be used within a session to distinguish between different packet flows. A single stream has a single reassembly queue and SHOULD be processed by a single core. Packets in a stream SHOULD NOT be distributed over multiple SCION paths to avoid reordering and therefore lower performance.
+The ingress SIG initiates unidirectional packet flows with the egress SIG simply by sending the corresponding SIG frames. There is no peamble. The egress SIG, should it accept the traffic, instanciates the necessary resources on-demand to process each flow. Each such flow forms an independent sequence of packets (a stream) ordered by an incrementing sequence number. Between a given SIG ingress/egress pair, a (session ID, stream ID) pair uniquely identifies a stream.
+
+To preserve performance, IP packets encapsulated in a single stream SHOULD NOT be re-ordered. To that end:
+
+- The ingress SIG SHOULD encapsulate IP packets that cannot be proven independent (for example with the same source, destination and traffic class) in the same stream.
+- The ingress SIG SHOULD encapsulate IP packets to a given stream in the order in which they were received.
+- The ingress SIG SHOULD send all frames of a given stream over a the same SCION path.
+- The egress SIG SHOULD reassemble and forward packets from each stream, ordered by frame sequence number and packet within each frame.
 
-Each SIG frame has a sequence number that is used by the egress SIG to reassemble the encapsulated IP packets within a stream.
+The session ID part of the (session ID, stream ID) pair is used to indicate traffic priority grouping. The egress SIG MAY dedicate processing resources to each session rather than to each individual stream.
```

Comment:
Actually the grouping does not guarantee any effective priority. The egress SIG has no idea what the various sessions are for. All it does is to give each session its own task, thereby preventing head-of-line blocking. The priorities arise from the relative loads placed by the sender on the sessions. In the OS impl, streams within a session have separate queues but they do not have separate tasks. That is the *only* effect of sessions that I could find in the impl, so I thought I'd use it as a way to explain why the protocol has an explicit ID pair to identify a stream: it enables this kind of hint from the ingress to the egress stream. Depending on implementation it may have an impact (e.g. in the OS impl) or not (e.g. if the egress SIG gives a core to every stream or has a single core running them all). This is basically the protocol enabling some subtle QoS behaviours without being too complicated. I am not sure how to explain that. If I don't people will ask what the two-level stream ID is for. I agree that "priority grouping" is crappy. Suggestions welcome for an alternate term.